text/0098-store-multiple-replay-segments-in-a-single-blob.md

```diff
@@ -0,0 +1,208 @@
+
+1. Click Tracking.
+   - Click events are published to the replay-event Kafka consumer.
+   - This publishing step is asynchronous and relies on threading to free up the main process thread.
+   - This operation is measured in microseconds and is not anticipated to significantly impact total throughput.
+2. Outcome Tracking.
+   - Outcome events are published to the outcomes Kafka consumer.
+   - This publishing step is asynchronous and relies on threading to free up the main process thread.
+   - This operation only occurs for segment-0 events.
+   - This operation is measured in microseconds and is not anticipated to significantly impact total throughput.
+3. Project lookup.
+   - Projects are retrieved by a cache lookup or querying PostgreSQL if it could not be found.
+   - This operation typically takes >1ms to complete.
+   - This operation only occurs for segment-0 events.
+   - Querying this information in a tight loop is not an ideal situation.
+     - Forwarding the project_id to a secondary Kafka consumer would free up resources on our main consumer and allow the secondary consumer to optimize for this type of workload.
+     - Alternatively, another method for looking up the project's `has_replay` flag could be found.
+
+**Duplicate Message Handling**
+
+1. Google Cloud Storage.
+   - Unique filename generation per buffer would mean that a segment could be present in multiple files.
+   - This has COGS implications but does not impact our application.
+2. "recording_byte_range" table.
+   - Duplicate replay, segment ID pairs will be recorded in the table.
+   - A reader must either select distinct or group by the replay_id, segment_id pair.
+   - Neither row has precendence over the other but the filename value must come from the same row as the start and stop byte range values.
+3. Outcome tracking.
+   - Duplicate outcomes will be recorded for a given replay.
+   - The replay_id functions as an idempotency token in the outcomes consumer and prevents the customer from being charged for the same replay multiple times.
+4. Click tracking.
+   - Duplicate click events will be inserted for a replay, segment pair.
+   - This is an acceptable outcome and will not impact search behavior.
+
+## Security
+
+How do we prevent users from seeing segments that do not belong to them? The short-answer is test coverage. Recording another segment's byte range would be the same as generating the incorrect filename under the current system. These outcomes are prevented with robust test coverage.
+
+In the case of bad byte range math, we do have some implicit protection. Each segment is compressed independently. Fetching a malformed byte range would yield unreadable data. A bad byte range either truncates the compression headers or includes unintelligible bytes at the beginning or end of the sequence. If we manage to decompress a subset of a valid byte range the decompressed output would be malformed JSON and would not be returnable to the user.
```

Comment:
What if the byte range accidentally includes a full other segment? (unlikely but worth considering imo)

Reply:
Then it would download that segment.  But that would be akin to a segment pointing to the wrong file name.  So its not a new failure mode.  With good test coverage neither of those cases should be possible.

Reply:
With the recent updates to the RFC regarding envelope encryption -- I would consider this mitigated. 👍

If bad byte range math happens, the extra bytes would not be decrypt-able without the matching Data Encryption Key.

Do we know how the system would behave if it were to be fed random bytes?

Reply:
The system doesn't exist yet but presumably it would error.  If we were persistently writing bad byte ranges (buggy ingestion process) then the replays would be lost.

Reply:
Also thanks @bmckerry for bringing this up and @mdtro for the great cryptography advice!  Team work :D